### ISSUE TITLE
Domain index entries could be generated later in the build process

### ISSUE THREAD

**Issue by cblegare** [sphinx-doc/sphinx#10299]
**Is your feature request related to a problem? Please describe.**

Domain index entries could be generated later in the build process.

**Describe the solution you'd like**

Domain indices, at least in HTML-based builders, are written during `Builder.finish` but generated during `Builder.prepare_writing`.  In other words

```
# Abstract builder when applicable, else HTML builder
Builder.build()
├── ...
├── Builder.write()
│   └── Builder.prepare_writing()
│       └── Index.generate()                # Here domain index entries are generated.
└── Builder.finish()
    └── Builder.gen_indices()
        └── Builder.write_domain_indices()  # Here domain index entries are used.
```

During `prepare_writing`, some class attributes from each domain Index are used, but not the generated content.  In other words, from what I have seen, the generated index entries could be generated (`Index.generate()`) directly during `Builder.write_domain_indices()`). In my opinion, it would make perfect sense to wait until the generated domain index entries are needed before generating them.  

Also, this enable domains to provide a wider variety of features along with their Index.

**Describe alternatives you've considered**

- https://github.com/sphinx-doc/sphinx/issues/10295
- Adding a finish task, but there is no API for this.  I would need to implement a custom builder.

**Additional context**

- [sphinx-dev post](https://groups.google.com/g/sphinx-dev/c/GaVMIwOA0sg)

---

In case you would consider this acceptable, I would probably write a patch since the required changes in the code seams quite minor and safe at first sight

**Comment by tk0miya** [comment]
The domain index entries are used to create rellinks on the navbar. So it's needed to generate before writing action, if my understanding is correct.
https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/builders/html/__init__.py#L481-L485

**Comment by cblegare** [comment]
Hello @tk0miya, this is exactly what I was referring to with the following comment:

> some class attributes from each domain Index are used, but not the generated content

Demonstration is below (under section **Index entries are not really used for rellinks**)

### tl;dr

- Simplest solution (**number 4**) involve to change `StandaloneHTMLBuilder.domain_indices` type and/or behavior, which might be OK if it is not considered as part of the public API.
  - before: `List[Tuple[str, Type[Index], List[Tuple[str, List[IndexEntry]]]`
  - after (possible) `List[Tuple[str, Index]]`
- A more compatible solution involves either a bit less readable and/or more duplicated code (**number 2**) or a quite big performance drop (**number 3**).
- Add a core event would provide *me* with an acceptable workaround (an be generally useful in my opinion) but would not fix the issue itself (**number 5**).

Please have a look below, especially at **number 2 and 4** (and 5 as a bonus).  I you find something suitable, I would submit a PR shortly.


### Index entries are not really used for rellinks

Following your comment

> The domain index entries are used to create rellinks on the navbar. So it's needed to generate before writing action, if my understanding is correct.

Yes and no.  Entries are used to exclude a domain index from rellinks when... the index generated no entries.  I've annotated the code snippet:

```python
#   ↓ used     ↓ used    ↓ not     ↓ not
for indexname, indexcls, _content, _collapse in self.domain_indices: 
    # if it has a short name 
    if indexcls.shortname:          # ← use of indexcls
        rellinks.append(
            (
                indexname,          # ← use of indexname
                indexcls.localname, # ← use of indexcls
                '',
                indexcls.shortname  # ← use of indexcls
            )
        ) 
```

Both strings `_content` and `_collapse` are only present on line 481 (this line) in this file.

the `domain_indices` attribute is written `prepare_writing`

https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/builders/html/__init__.py#L440-L455

**but** a domain index is **not** added to `domain_indices` (and thus to rellinks) if it generates no entries.

### Challenges for fixing this

- If the generated index `content` is falsy, the index is not added at all to `domain_indices` (line 453), thus not added to `rellinks`.
- the `domain_indices` attribute is not in the `Builder` API, but it is still a public attribute. Is is present in all builder extending the `StandaloneHTMLBuilder` and I've also found it in the LaTeX builder, but its use is slightly different.
- For the EPub builder, all solutions proposed below would be compatible
  https://github.com/sphinx-doc/sphinx/blob/b12b39db05d605cfb946f84afd27e0f6cf3c9e13/sphinx/builders/_epub_base.py#L542-L545


### Some possible solutions

Here are the options I see here

#### 1. Bad idea, keep as-is

That would be sad, but all right

#### 2.  Don't check if the index generates anything before adding it to `rellinks`

which boils done to changing these 

https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/builders/html/__init__.py#L452-L455
with

```python
self.domain_indices.append((indexname, indexcls, [], False))
```

then in 
    https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/builders/html/__init__.py#L709-L717

add code to find back the domain, make the index instance, then generate

It would also be possible to remove the rellinks from within `write_domain_indices` 
if they generate nothing, since it happens before the rellinks are used.

Pros/Cons in my opinion:
- *pros*
  - All attributes behave the same (no compatibility break)
- *cons*
  - Complicated code for `write_domain_indices`
  - Possible rellinks to empty domain indices unless removed in `write_domain_indices`

#### 3.  Use `Index.generate` twice

Ignore the generated content in `prepare_writing` and generate it again in `write_domain_indices` (performance drop)

- *pros*
  - Everything behave exactly the same
- *cons*
  - Performance drop because domain index entries are generated twice

I suspect this is simply not acceptable.


#### 4.  Do not consider `domain_indices` to be in the API

Change its behavior and type.

- change `domain_indices` to store indexname and index (an instance, not a class)
- Don't check if the index generating anything before adding it to `domain_indices`
- Add rellinks for indices that are defined in config even if they generate nothing
  - It would also be possible to remove the rellinks from within `write_domain_indices` 
    if they generate nothing, since it happens before the rellinks are used.

- *pros*
  - Simple, does the job
- *cons*
  - Change the `domain_indices` attribute API for users.  From within the Sphinx codebase, it looks quite safe

#### 5. Introduce some hooks early in the finishing tasks

The `Builder.finish` API looks... unfinished ;)

- The default implementation does nothing (up to the builder)
  https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/builders/__init__.py#L581-L586
- It is called at the end of `Builder.build` after all the *regular* files have been written
  https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/builders/__init__.py#L350-L356
- It seems tied to the `finish_tasks` attribute, but the `Builder` abtractclass itself does nothing with it
- The SerialTask and ParallelTask (type of `Builder.finish_tasks`) will execute a task immediately when it is added, so it is not possible to `self.env.builder.finish_tasks.add_task` earlier in the build process (when there are events to hook on that are not specific to a docname)
  https://github.com/sphinx-doc/sphinx/blob/4221d1a5163d023d80a34bd4a12e9c2a2c648ae3/sphinx/util/parallel.py

The only event that happens after individual docnames 
have been written is `build-finished` which is quite late.

- *pros*
  - that kind of event could be quite useful!
  - I would suffer a performance drop if I would use this even 
    to re-generate index entries, my library only, not the whole Sphinx community
  - Fully backward compatible
- *cons*
  - does not exactly fix the issue

**Comment by cblegare** [comment]
Hello again @tk0miya

The above PR implements possible solution number **4 Do not consider domain_indices to be in the API** from the comment above: https://github.com/sphinx-doc/sphinx/issues/10299#issuecomment-1085341741
